File: demo/pfizer_audit_logger.go
/*
to build: install go tools,
> go get github.com/mssola/user_agent
> go build pfizer_audit_logger.go

This script will read json-encoded auditlog messages from Transmart and send them as HTTP POST messages to
metrics.pfizer.com. The input must be provided as json objects on stdin. If a request fails this is noted
on stderr, but the request is not retried so in that case the message is lost. If more than MAX_QUEUE_LENGTH
http operations are in progress at the same time any further input messages are ignored until some http
operations finish.
*/

package main

import (
	"os"
	"time"
	"sync"
	"sync/atomic"
	"encoding/json"
	"fmt"
	"strings"
	"net/url"
	"net/http"
	"io/ioutil"
	"github.com/mssola/user_agent"
)

const URL = "http://metrics.pfizer.com/metrics/servlet/RegisterEventServlet"

// If there are more than this number of unprocessed messages, further messages will be dropped until the server
// catches up.
const MAX_QUEUE_LENGTH = 1000


var wg sync.WaitGroup
// The internal count of WaitGroup is not accessible, so we maintain another copy to know how many requests are in progress
var requests_in_progress int32 = 0

func atomicAdd(i *int32, delta int32) { atomic.AddInt32(i, delta) }
func atomicLoad(i *int32) int32 { return atomic.LoadInt32(i) }


/* struct based implementation */
type jsonMsg struct {
       Action string
       Study string
       Subset1 string
       Subset2 string
       Application string `json:"program"`
       AppVersion string `json:"programVersion"`
       User string
       Task string `json:"event"`
       Browser string `json:"userAgent"`
}
/**/

/* Alternative map based implementation 
type jsonMsg map[string]string

var keyMap = map[string]string{
	"program": "application",
	"programVersion": "appVersion",
	"event": "task",
}
*/


func send_auditlog_record(msg jsonMsg) {

	//fmt.Println(msg)
	params := url.Values{}

/* struct based implementation */

	if msg.Action != "" {
		params.Set("action", msg.Action)
	} else {
		action := []string{}
		for _, s := range []string{msg.Study, msg.Subset1, msg.Subset2} {
			if s != "0" && s != "" {
				action = append(action, s)
			}
		}
		params.Set("action", strings.Join(action, "|"))
	}
	if msg.Application != "" { params.Set("application", msg.Application) }
	if msg.AppVersion != "" { params.Set("appVersion", msg.AppVersion) }
	if msg.User != "" { params.Set("user", msg.User) }
	if msg.Task != "" { params.Set("task", msg.Task) }
	if msg.Browser != "" {
		browser, version := user_agent.New(msg.Browser).Browser()
		params.Set("browser", browser + " " + version)
	}
/**/

/* alternative map based implementation 

	for _, k := range []string{"action", "program", "programVersion", "user", "event"} {
		if _, hasKey := msg[k]; !hasKey { continue }
		key := k
		if outkey, translate := keyMap[k]; translate {
			key = outkey
		}
		params.Set(key, fmt.Sprintf("%v", msg[k]))
	}
	if _, hasKey := msg["action"]; !hasKey {
		action := []string{}
		for _, s := range []string{msg["study"], msg["subset1"], msg["subset2"]} {
			if s != "0" && s != "" {
				action = append(action, s)
			}
		}
		params.Set("action", strings.Join(action, "|"))
	}
	if _, hasKey := msg["userAgent"]; hasKey {
		browser, version := user_agent.New(msg["userAgent"]).Browser()
		params.Set("browser", browser + " " + version)
	}
*/

	fullurl := URL + "?" + params.Encode()

	//fmt.Println(fullurl)
	resp, err := http.PostForm(fullurl, url.Values{})
	if err != nil {
		error("WARNING: Failed to post message to metrics server: "+err.Error())
	} else {
		ioutil.ReadAll(resp.Body)
		resp.Body.Close()
	}

	atomicAdd(&requests_in_progress, -1)
	wg.Done()
}


func error(m string) {
	fmt.Fprintln(os.Stderr, time.Now().Format("2006-01-02 15:04:05.000-07:00  ") + m)
}

func fatal(m string) {
	error(m)
	os.Exit(1)
}


func main() {
	
	dec := json.NewDecoder(os.Stdin)
	msg := jsonMsg{}

	for dec.More() {
		err := dec.Decode(&msg)
		if err != nil { fatal("ERROR: Failed to parse JSON message from stdin: " + err.Error()) }
		if atomicLoad(&requests_in_progress) > MAX_QUEUE_LENGTH {
			line, _ := json.Marshal(msg)
			error("WARNING: Too many unfinished requests in progress, ignoring message "+string(line))
			continue
		}
		atomicAdd(&requests_in_progress, 1)
		wg.Add(1)
		go send_auditlog_record(msg)
		msg = jsonMsg{}
	}

	//fmt.Println("done reading")
	wg.Wait()
	//fmt.Println("Exiting")
}

File: demo/pfizer_audit_logger.py
#!/usr/bin/env python3
#
# Tested with python 3.4
#
# This script will read json-encoded auditlog messages from Transmart and send them as HTTP POST messages to
# metrics.pfizer.com. The input must be provided as json objects separated by newlines. The json objects themselves
# must not contain unencoded newlines.
#
# This script will make up to THREADS requests in parallel to the metrics server. If a request fails this is noted
# on stderr, but the request is not retried so in that case the message is lost.

import sys, os
import logging
import concurrent.futures
import json
from threading import Lock
import urllib.request
import urllib.parse
from urllib.request import Request
# httpagentparser is not in the standard library, use "pip3 install httpagentparser" to install it
import httpagentparser

URL = 'http://metrics.pfizer.com/metrics/servlet/RegisterEventServlet'
# Maximum timeout when waiting on the metrics server
TIMEOUT = 10 #seconds
# If there are more than this number of unprocessed messages, further messages will be dropped until the worker
# threads catch up.
MAX_QUEUE_LENGTH = 1000
# Max number of parallel requests to the metrics server
THREADS = 20


countlock = Lock()
queuelength = 0
failcount = 0

logging.basicConfig(level=logging.INFO, format='{asctime} {levelname}: {message}', style='{',
                    datefmt="%Y-%m-%d %H:%M:%S%z")
log = logging.getLogger()

def send_auditlog_record(line):
    global failcount, queuelength

    with countlock:
        queuelength -= 1

    # Invalid json input is not an error we can handle
    msg = json.loads(line)
    task = msg['event']
    user = msg['user']
    if task == "User Access":
        action = user
    else:
        action = msg.get('action') or '|'.join(msg.get(x) for x in ('study', 'subset1', 'subset2', 'analysis', 'query', 'facetQuery', 'clientId') if msg.get(x))
    args = dict(action = action,
                application = msg['program'],
                appVersion = msg['programVersion'],
                user = user,
                task = task,
            )
    if msg['userAgent']:
        args['browser'] = '<unknown browser>'
        browser = httpagentparser.detect(msg['userAgent']).get('browser')
        if browser:
            args['browser'] = browser['name'] + ' ' + browser['version']
    fullurl = URL + '?' + urllib.parse.urlencode(args)
    #print(fullurl)
    try:
        #raise Exception('testing')
        urllib.request.urlopen(Request(fullurl, method='POST'), timeout=TIMEOUT).readall()
    except Exception as exc:
        with countlock:
            failcount += 1
        log.error("{e}, url: {url}".format(e=' '.join(str(e) for e in exc.args), url=fullurl))

def process(line):
    try:
        send_auditlog_record(line)
    except BaseException as e:
        # An exception here is a programming error or something serious. As this is not the main thread we can't just
        # let the exception bubble up, so kill ourselves forcefully.
        log.fatal(str(e) + ", aborting!", exc_info=e)
        os.abort()


with concurrent.futures.ThreadPoolExecutor(max_workers=THREADS) as executor:
    for line in sys.stdin:
        if queuelength > MAX_QUEUE_LENGTH:
            log.error("MAX_QUEUE_LENGTH exceeded, ignoring message {line}".format(line=line))
            continue
        with countlock:
            queuelength += 1
        executor.submit(process, line)

sys.stdin.close()
